lib/puppet-lint.rb

```diff
@@ -175,14 +175,16 @@ def report(problems)
 
       next unless message[:kind] == :fixed || [message[:kind], :all].include?(configuration.error_level)
 
-      if configuration.json || configuration.sarif
+      if configuration.json || configuration.sarif || configuration.codeclimate_report_file
         message['context'] = get_context(message) if configuration.with_context
         json << message
-      else
-        format_message(message)
-        print_context(message) if configuration.with_context
-        print_github_annotation(message) if configuration.github_actions
       end
+
+      next if configuration.json || configuration.sarif
```

Comment:
I'm never a fan of `next`. Perhaps a much simpler API is to always return an array of problems, regardless of the mode. So really simplified:

```ruby

result = []

print_stdout = !(configuration.json || configuration.sarif || configuration.codeclimate_report_file)

problems.each do |message|
  # kind mangling code here

  result << message

  if print_stdout
    format_message(message)
    print_context(message) if configuration.with_context
    print_github_annotation(message) if configuration.github_actions
  else
    message['context'] = get_context(message) if configuration.with_context
  end
end

result
```

Then I should also note that GitHub is deprecating the printing of output to stdout, but now recommend to append to `GITHUB_OUTPUT`. See https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/ as well. This means we can easily also run it safely, regardless of output mode. Separate PR, but it can make the flow of this easier.

Then I'd also ask if setting `message['context']` is really something we need to guard behind `print_stdout`. Instead, perhaps we can modify `print_context` to utilize `message['context']` instead.

Sorry, lots of unrelated suggestions, but this report method is getting very complicated.

Reply:
So something along the lines of https://github.com/puppetlabs/puppet-lint/pull/83.

Reply:
I think I've correctly cherry-picked your refactoring suggestions.

Reply:
I would have expected a rebase instead of follow ups, but I guess both work.